### ISSUE TITLE
Internal Error: No cycle found with nodes with common side effects on build failure

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2777]
This issue was originally created at: 2011-07-20 07:29:37.
This issue was reported by: `estanchev`.
estanchev said at 2011-07-20 07:29:37
>(Attached is a diagram that might help to illustrate the problem)
> 
>If there are two nodes that share the same side effect and the first one is being executed, the other one can become a pending child of a third node while waiting on the side effect.
>
>If the first node fails to build, it will clear the pending child set for all of its parents, but the other "branch" in the DAG containing the second node will not be cleared, causing this type of error:
> 
```
UserError: Found dependency cycle(s):
  Internal Error: no cycle found for node n2 (<__main__.Node object at 0xec8810>) in state pending
```

>The proposed patch clears the pending children from all children waiting on the side effect.
>
>At first, I thought I could just add the `waiting_s_e` to `to_visit` and prune `waiting_s_e` of the side effect, but I think that setting the state of the parents of the side effect to 'failed' is incorrect.
> 
>I've written a simple testcase in the taskmaster tests, which illustrates the problem. With unpatched taskmaster it fails with the above error. With the patch applied, it succeeds.
> 
>PS: patch is based on the github clone of the scons repo, I hope that's fine. It was just easier for me to use that.

estanchev said at 2011-07-20 07:30:52
>Created an attachment (id=858)
>This illustrates the problem. If n1 fails, n2 and its parents won't be removed from the pending children set

estanchev said at 2011-07-20 07:32:06
>Created an attachment (id=859)
>This is the proposed patch.

estanchev said at 2011-08-31 06:11:41
>Created an attachment (id=867)
>Updated patch: forgot to decrement ref_counts!

garyo said at 2012-06-14 09:48:01
>In the second patch (id=867) nothing defines the `get_side_effects` method on a Node.  It looks to me like the second patch is a bit damaged.  The first patch (id=859) adds the Node methods properly.  The second one (which is supposed to be an updated version of the first) is missing those changes.  I haven't reviewed in detail but it looks like there's a bit more work to do on this.  Emil, if you're reading this, could you take a look?  Leave a note here if you can, otherwise someone else will try to handle it.

estanchev said at 2012-06-14 12:29:29
>Created an attachment (id=894)
>Fixing patch v2, which I apparently created concatenating patches for each git commit instead of a single patch. Sorry about that.

evgen said at 2016-07-07 17:32:35
>Does this only happen with SideEffect?

rzr said at 2017-07-10 07:22:33
>Any update on this issue ? it would be nice to have it fixed in next release.

rzr said at 2017-08-27 07:36:05
>As suggested my upstream on IRC, I created this pull request:
>https://bitbucket.org/scons/scons/pull-requests/508/fix-a-pending-children-and-side-effects/diff

rzr said at 2017-10-10 02:32:36
>Also under review at:
>https://github.com/SConsProject/scons/pull/11

Votes for this issue: 15.

estanchev attached [scons-pending.png](https://github.com/SCons/tigris-issue-attachments/blob/master/2777/858/scons-pending.png) at 2011-07-20 07:30:52.
>This illustrates the problem. If n1 fails, n2 and its parents won't be removed from the pending children set

estanchev attached [Fix-a-pending-children-and-side-effects-bug.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2777/859/Fix-a-pending-children-and-side-effects-bug.patch) at 2011-07-20 07:32:06.
>This is the proposed patch.

estanchev attached [Fix-a-pending-children-and-side-effects-bug_v2.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2777/867/Fix-a-pending-children-and-side-effects-bug_v2.patch) at 2011-08-31 06:11:41.
>Updated patch: forgot to decrement ref_counts!

estanchev attached [Fix-a-pending-children-and-side-effects-bug_v3.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/2777/894/Fix-a-pending-children-and-side-effects-bug_v3.patch) at 2012-06-14 12:29:29.
>Fixing patch v2, which I apparently created concatenating patches for each git commit instead of a single patch. Sorry about that.
